src/middlewared/middlewared/plugins/vm.py

```diff
@@ -1634,6 +1654,69 @@ def initialize_vms(self, timeout=10):
                 # Whatever happens, we don't want middlewared not booting
                 self.vms[vm_data['name']] = VMSupervisor(vm_data, self.libvirt_connection, self.middleware)
 
+    @private
+    def setup_libvirt_events(self):
+        def callback(conn, dom, event, detail, opaque):
+            """
+            0: 'DEFINED',
+            1: 'UNDEFINED',
+            2: 'STARTED',
+            3: 'SUSPENDED',
+            4: 'RESUMED',
+            5: 'STOPPED',
+            6: 'SHUTDOWN',
+            7: 'PMSUSPENDED'
+            Above is event mapping for internal reference
+            """
+            vms = {VMSupervisor.domain_name(d): d for d in self.middleware.call_sync('vm.query')}
+            if dom.name() not in vms:
+                emit_type = 'REMOVED'
+            elif event == 0:
+                emit_type = 'ADDED'
+            else:
+                emit_type = 'CHANGED'
+
+            vm_state_mapping = {
+                0: 'NOSTATE',
+                1: 'RUNNING',
+                2: 'BLOCKED',
+                3: 'PAUSED',
+                4: 'SHUTDOWN',
+                5: 'SHUTOFF',
+                6: 'CRASHED',
+                7: 'PMSUSPENDED',
+            }
+            try:
+                if event == 1:
+                    if emit_type == 'REMOVED':
+                        state = 'NOSTATE'
+                    else:
+                        # We undefine/define domain numerous times based on if vm has any new changes
+                        # registered, this is going to reflect that
+                        state = 'UPDATING CONFIGURATION'
+                else:
+                    state = vm_state_mapping.get(dom.state()[0], 'UNKNOWN')
+            except libvirt.libvirtError:
+                state = 'UNKNOWN'
+
+            if dom.name().split('_')[0].isdigit():
```

Comment:
We are doing this name() split twice. Just a nitpick but we should use a var instead